linked_list/linked_list.py

```diff
@@ -16,64 +19,152 @@ def __init__(self):
+                return True
+            current = current.next
+        
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity:    O(1)
     def length(self):
-        pass
+        if not self.head:
+            return 0
+        
+        current = self.head
+        count = 0
+        while current:
+            count += 1
+            current = current.next
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity:   o(1)
     def get_at_index(self, index):
-        pass
+        if self.length() <= index:
+            return None
+
+        current = self.head
+        count = 0
+
+        while current:
+            if count == index:
+                return current.value
+            count += 1
+            current = current.next
+        return None 
+
+        
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def get_last(self):
-        pass
+        if not self.head:
+            return None
+        current = self.head 
+        while current.next:
+            current = current.next
+        return current.value 
+        # return self.tail.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.add_first(value)
+        else: 
+            new_node = Node(value)
+            current = self.head
+            while current.next: 
+                current = current.next
+            current.next = new_node 
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if not self.head: 
+            return None 
+        current = self.head
+        max = current.value
+
+        while current: 
+            if max < current.value: 
+                max = current.value 
+            current = current.next
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)
+    # Space Complexity: o(1)
     def delete(self, value):
-        pass
+        if not self.head: 
+            return 
+        # current = self.head 
+        elif current.value == value:
+            self.head = current.next
+            return True 

```

Comment:
You don't need to return a value here
```suggestion
            return

```